Throwing my hat in the ring for how to handle pathnames.

indifferentghost opened the pull request:
This is a suggestion of how to handle pathnames in mock-url #6.

Originally I had the pathname as an input field, but if the [can-route-pushstate.root](https://canjs.com/doc/can-route-pushstate.prototype.root.html) was set it would _ever so often_ reset to the root.

Demo: https://codepen.io/TDillard/pen/WaPmKw?editors=0010

indifferentghost committed:
updated with suggestion of how to handle pathnames. This is namely an update for handling examples in can-route-pushstate.

indifferentghost committed:
removed code that was an artifact remaining when path was two-way-bound to an input element.

indifferentghost committed:
made pushstate toggleable.

indifferentghost committed:
spaces to tabs

indifferentghost commented:
It's now configurable!
```html
<mock-url pushstate:from="true"></mock-url>
```
Enables it to catch [pushstate](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Example_of_pushState()_method).

phillipskevin commented on mock-url.mjs:
```diff
@@ -1,98 +1,129 @@
-
-import { Component, DefineMap, stache } from "//unpkg.com/can@5/core.mjs";
+import { Component, DefineMap, stache } from "can";
```
Removing the url for `can` breaks the demo page.

phillipskevin commented on mock-url.mjs:
```diff
@@ -1,98 +1,129 @@
+			<span class='reload' on:click='reload()'>&#8635;</span>
+			<div class="url">
+				<span class='base'>URL:{{page}}{{path}}</span>
+				<span class='input'><input value:bind="url" placeholder="#! The hash is empty"/></span>
```
This doesn't seem to work correctly if using pushstate. The `path` should be editable.

phillipskevin commented on mock-url.mjs:
```diff
@@ -1,98 +1,129 @@
+			default: "/my-app.html"
+		},
+		pushstate: {
+			default: false
```
Give this a `type: "boolean"` so that `:raw` or `:from` bindings will both work

phillipskevin commented on mock-url.mjs:
```diff
@@ -1,98 +1,129 @@
+				};
+				
+				// creating default pathname
+				window.history.replaceState( null, null, '/' );
```
If using a `root` other than `/`, won't this cause the URL to navigate away from the app?

indifferentghost replied:
It'll only navigate if the `location.pathname` is directly edited, as far as I've been able to find.

phillipskevin reviewed (dismissed):
I couldn't get this to work with an app that is using pushstate. I'll take another look once some of the requested changes are fixed up.

indifferentghost committed:
path now is editable, pushstate type is boolean.

indifferentghost committed:
fixed imports.

indifferentghost committed:
Update package.json

indifferentghost committed:
Reintroduced pushState, added onpopstate event fn

- `window.history.pushState` is no longer a copy of replaceState. It now
is calls the native pushState function. This change was made so browser
API navigation (ie. `history.forward()`, `history.back()`) will work.

- Added `window.onpopstate` event handler to update the `path`
observable when API navigation functions are called.

indifferentghost committed:
5.1.0

indifferentghost committed:
Pushstate OR hash, general UI improvements

- rendering urls based on pushstate or hash for immersion
- removed preceeding "URL:"
- removed .path class
- removed reload button

indifferentghost committed:
fixed bug with adding text while root is set.

indifferentghost committed:
Fixed "root", pushState now works with hashes
